[C++] Fix GLPK bound type + numerical issues

jmirabel opened the pull request:
Fixes #

Changes in this PRs:
- 
- 

Checklists:
- [x] Update CHANGELOG.md with a single line describing this PR

jmirabel commented:
I use https://cli.github.com/ and, as I discover it now, it does not follow PR templates...

hungpham2511 commented:
No worry. Thanks for the PR. Could you help to add a line to HISTORY.md? Then feel free to merge whenever you are happy.

jmirabel commented:
Do you know why two CI are not run and do not have a *Details* link ?

hungpham2511 commented:
I wonder that as well. Probably because I renamed the workflows in the last PRs.

jmirabel commented on cpp/src/toppra/solver/glpk-wrapper.cpp:
```diff
@@ -13,20 +13,26 @@ void intersection (const Bound& a, const Bound& b, Bound& c)
+    if (u == infty) return GLP_FR;
+    else return GLP_UP;
+  } else {
+    if (l >= u) return GLP_FX;
```
The `>=` fixes #112 for GLPK.

hungpham2511 commented on cpp/src/toppra/solver/qpOASES-wrapper.cpp:
```diff
@@ -118,6 +118,9 @@ bool qpOASESWrapper::solveStagewiseOptim(std::size_t i,
   int nWSR = 1000;
+  // Make sure bounds are correct
+  h = h.cwiseMax(l);
+  m_hA = m_hA.cwiseMax(m_lA);
```
Interesting! I wonder if this fixes the issue, or it actually a bug. Why would h not larger than l? And m_hA not larger than m_lA? From the constraint evaluation rule, this condition should always held.

hungpham2511 replied:
My understanding of the fowardpass-numerical issue is that the controllable sets condition and the dynamics conditions barely become infeasible.  Now, these are in separate groups of rows of the A matrix, so taking the component-wise max/min should not fix this issue.

jmirabel replied:
> Why would h not larger than l? And m_hA not larger than m_lA?

In the case that concerned me, some h coeffs were smaller than the corresponding l by a very tiny coefficient (see values in #112). Commenting out m_hA update does not introduce errors so the only useful line in my case is `h = h.cwiseMax(l)`.

hungpham2511 replied:
This works for me in the instance that I had issue with. Tested with qpoases only.

```
diff --git a/cpp/src/toppra/algorithm.cpp b/cpp/src/toppra/algorithm.cpp
index b12e476..c855c4c 100644
--- a/cpp/src/toppra/algorithm.cpp
+++ b/cpp/src/toppra/algorithm.cpp
@@ -3,6 +3,7 @@
 #include <cstddef>
 #include <iostream>
 #include "toppra/toppra.hpp"
+#include <algorithm>

 namespace toppra {

@@ -55,14 +56,16 @@ ReturnCode PathParametrizationAlgorithm::computeControllableSets(
     solver_ret = m_solver->solveStagewiseOptim(i, H, g_lower, x, x_next, solution);

     TOPPRA_LOG_DEBUG("down: " << solution);
-
     if (!solver_ret) {
       ret = ReturnCode::ERR_FAIL_CONTROLLABLE;
       TOPPRA_LOG_DEBUG("Fail: controllable, lower problem, idx: " << i);
       break;
     }

-    m_data.controllable_sets(i, 0) = solution[1];
+    // For whatever reason, sometimes the solver return negative
+    // solution despite having a set of positve bounds. This readjust
+    // if the solution is negative.
+    m_data.controllable_sets(i, 0) = std::max(0.0, solution[1]);
   }
   return ret;
 }
```

jmirabel replied:
I tested with this change and it does not solve my issue. I still need `h = h.cwiseMax(l)`

jmirabel replied:
Do you want me to remove the `m_hA` update ?

hungpham2511 replied:
That's quite concerning actually. Are you able to provide the example you are having difficulty with? Though `h=h.cwiseMax(l)`, we might overlook a bug.

In fact, I think both l/h and lA/hA constraints should be assertions instead.

jmirabel replied:
> Are you able to provide the example you are having difficulty with?

I can provide easily the position, velocity, acceleration and times together with the URDF model (for testing purpose, I use PR2).

hungpham2511 replied:
> > Are you able to provide the example you are having difficulty with?
> 
> I can provide easily the position, velocity, acceleration and times together with the URDF model (for testing purpose, I use PR2).

Alright, I propose merging #114 first,  then you provide the failing case in this PR and I can help to inspect this numerical issue/bug.

jmirabel replied:
Ok. Go ahead.

jmirabel replied:
While trying to generate the data, I realize that all the acceleration of my path are 0. Can this be my problem ?

jmirabel replied:
I realize now that the problem I try to solve is not pertinent. I was too focus on the software problem.

hungpham2511 commented:
Also, I think it's worth mentioning that you should use the INTERPOLATION constraint scheme. It provides significantly better numerical stability and solution quality.

jmirabel commented:
Same as #115, if you are ok with the changes, I propose to bypass the not-run CI and merge this.

jmirabel commented:
I updated this PR as follows:
- boundary issues are reported as failure (same behavior as before this PR, with some additional logs),
- I added a failing test case.
- this PR also contains a few minor unrelated bug fix.

hungpham2511 commented on integration_tests/failing_test.py:
```diff
@@ -0,0 +1,220 @@
+from toppra.algorithm import TOPPRA
+algo = TOPPRA(constraints, path, solver_wrapper='seidel', gridpoints=
+        np.linspace(0, path.duration, Ngripoints+1))
+sdd, sd, _ = algo.compute_parameterization(0, 0)
```
Just a comment, normally the number of grid points should be at least a few times the waypoint, especially if the path is not smooth in-between waypoints.

That said, definitely, the cpp implementation is having a numerical issue because the algorithm should not fail in the forward pass regardless of the number of waypoints. I will help to look into this.

jmirabel replied:
I noticed that in the readme, thanks for remembering it. My intent here was to give the same points to TOPPRA as in my case in order to replicate its evaluation.
I saved the values of q, v and a at each gridpoint. To replicate TOPPRA evaluation, I must give a path and gridpoints which will give the same q, v and a.

jmirabel replied:
> I will help to look into this.

Thank you for this.

hungpham2511 replied:
The problem is due to the way floats are treated in C++ Apparently 

```
      l[1] = std::max(l[1], box.x[i][0]);
      h[1] = std::min(h[1], box.x[i][1]);
```
Can lead to l[1] > h[1] even if x[i][0] <= x[i][1]. The difference is very small, similar to what you have found ealier. http://www.cplusplus.com/forum/beginner/235970/

I think we need to do what you have done earlier (using cwiseMax). To ensure that the constraints are correctly formed, we will need to put assertions in the preparation steps.

I made the change locally and manage to run your failing test. The velocity differences are ~ 1e-8. 

Note that the python instance needs to use interpolation scheme as well:

```
modified   integration_tests/failing_test.py
@@ -2,13 +2,18 @@ import pinocchio
 import numpy as np
 from toppra.cpp import Interpolation
 
+import toppra
+toppra.setup_logging(level="DEBUG")
+
 def torque_constraint(robot):
-    from toppra.constraint import JointTorqueConstraint
+    from toppra.constraint import JointTorqueConstraint, DiscretizationType
     def inv_dyn (q, v, a):
         return pinocchio.rnea(robot.model, robot.data, q, v, a)
-    return JointTorqueConstraint (inv_dyn,
+    c = JointTorqueConstraint (inv_dyn,
         np.vstack([-robot.model.effortLimit,robot.model.effortLimit]).T,
         np.zeros(robot.model.nv))
+    c.discretization_type = DiscretizationType.Interpolation
+    return c
 
```

jmirabel replied:
> The problem is due to the way floats are treated in C++ Apparently
> 
> ```
>       l[1] = std::max(l[1], box.x[i][0]);
>       h[1] = std::min(h[1], box.x[i][1]);
> ```
> 
> Can lead to l[1] > h[1] even if x[i][0] <= x[i][1]. The difference is very small, similar to what you have found ealier. http://www.cplusplus.com/forum/beginner/235970/

I don't think it can be that. Rounding errors in C++ are not different than in Python. Floating point comparison is transitive.

I checked that values in the place above for the current failing test. The value for `l[1]` comes from `x[0]` while `h[1]` comes from `box.x[2][1]`. So definitely, there is a constraint which is in conflicts with the search interval.

hungpham2511 replied:
It's possible that there are multiple box constraints that are conflicting, in which case the condition cwiseMax will create a bug.

Though, each i is treated separately, so box.x[0] wouldn't relate to box.x[2]. Did you have a typo?

hungpham2511 replied:
Ok I think I find out the actual cause, it's because `xNext` conflict with `box.x[i]`. Most likely there is a rounding error in the previous forward step.

jmirabel replied:
> Though, each i is treated separately, so box.x[0] wouldn't relate to box.x[2]. Did you have a typo?

No, it is not a typo. `x[0]` does not refer to `box.x[0]` but to the first value of the vector `x` passed as argument. On my computer, `x[0]` and `box.x[2][1]` conflicts. `x[0]` coming from the integration in the forward pass

    m_data.parametrization(i + 1) =
      std::max(0., m_data.parametrization(i) + 2 * deltas(i) * solution(0));

So I agree that the issue is a rounding error in this line.

hungpham2511 replied:
This diff works for me. Without the cwiseMin/Max, the assertions will fire. Look like this is the issue. 

```
@@ -150,6 +158,11 @@ bool qpOASESWrapper::solveStagewiseOptim(std::size_t i,
   if (res == qpOASES::SUCCESSFUL_RETURN) {
     solution.resize(nbVars());
     m_impl->qp.getPrimalSolution(solution.data());
+    TOPPRA_LOG_DEBUG("solution: " << solution);
+    solution = solution.cwiseMax(l.transpose());
+    solution = solution.cwiseMin(h.transpose());
+    assert((solution.transpose().array() <= h.array()).all());
+    assert((solution.transpose().array() >= l.array()).all());
     return true;
   }
```

jmirabel replied:
This looks like a good solution to me !

hungpham2511 replied:
I added this fix in #120.

jmirabel commented:
I updated and re-based this PR. Is there anything else I should do ?

hungpham2511 commented:
> I updated and re-based this PR. Is there anything else I should do ?

The PR looks good to me. If you are ok, feel free to merge after CI passes. Do you still have problem with the failing test?

jmirabel commented:
I may do the same thing in the GLPK solver later.
